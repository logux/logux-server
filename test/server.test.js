'use strict'

const spawn = require('cross-spawn')
const path = require('path')

const Server = require('../server')

const DATE = /\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d/g

let started

function start (name, args) {
  return new Promise(resolve => {
    started = spawn(path.join(__dirname, '/servers/', name), args)
    let running = false
    function callback () {
      if (!running) {
        running = true
        resolve()
      }
    }
    started.stdout.on('data', callback)
    started.stderr.on('data', callback)
  })
}

function check (name, args, kill) {
  return new Promise(resolve => {
    let out = ''
    const server = spawn(path.join(__dirname, '/servers/', name), args)
    server.stdout.on('data', chank => {
      out += chank
    })
    server.stderr.on('data', chank => {
      out += chank
    })
    server.on('close', exit => {
      let fixed = out
        .replace(DATE, '1970-01-01 00:00:00')
        .replace(/"time":"[^"]+"/g, '"time":"1970-01-01T00:00:00.000Z"')
        .replace(/PID:(\s+)\d+/, 'PID:$121384')
        .replace(/"pid":\d+,/g, '"pid":21384,')
        .replace(/Logux server:( +)\d+.\d+.\d+/g, 'Logux server:$10.0.0')
        .replace(/"loguxServer":"\d+.\d+.\d+"/g, '"loguxServer":"0.0.0"')
        .replace(/"hostname":"[^"]+"/g, '"hostname":"localhost"')
      fixed = fixed.replace(/\r\v/g, '\n')
      resolve([fixed, exit])
    })

    function waitOut () {
      if (out.length > 0) {
        server.kill('SIGINT')
      } else {
        setTimeout(waitOut, 100)
      }
    }
    if (kill) setTimeout(waitOut, 200)
  })
}

function checkOut (name, args) {
  return check(name, args, 'kill').then(result => {
    const out = result[0]
    const exit = result[1]
    if (exit !== 0) {
      throw new Error(`Fall with:\n${ out }`)
    }
    expect(out).toMatchSnapshot()
  })
}

function checkError (name, args) {
  return check(name, args).then(result => {
    const out = result[0]
    const exit = result[1]
    expect(exit).toEqual(1)
    expect(out).toMatchSnapshot()
  })
}

const originEnv = process.env.NODE_ENV
afterEach(() => {
  process.env.NODE_ENV = originEnv
  delete process.env.LOGUX_PORT
  delete process.env.LOGUX_REPORTER
  if (started) {
    started.kill('SIGINT')
    started = undefined
  }
})

it('uses CLI args for options', () => {
  const options = Server.loadOptions({
    argv: [
      '',
      '--port', '31337',
      '--host', '192.168.1.1',
      '--reporter', 'json'
    ],
    env: { }
  })

  expect(options.host).toEqual('192.168.1.1')
  expect(options.port).toEqual(31337)
  expect(options.reporter).toEqual('json')
  expect(options.cert).toBeUndefined()
  expect(options.key).toBeUndefined()
})

it('uses env for options', () => {
  const options = Server.loadOptions({
    argv: [],
    env: {
      LOGUX_HOST: '127.0.1.1',
      LOGUX_PORT: 31337,
      LOGUX_REPORTER: 'json'
    }
  })

  expect(options.host).toEqual('127.0.1.1')
  expect(options.port).toEqual(31337)
  expect(options.reporter).toEqual('json')
})

it('uses combined options', () => {
  const options = Server.loadOptions({
    env: { LOGUX_CERT: './cert.pem' },
    argv: ['', '--key', './key.pem']
  }, { port: 31337 })

  expect(options.port).toEqual(31337)
  expect(options.cert).toEqual('./cert.pem')
  expect(options.key).toEqual('./key.pem')
})

it('uses arg, env, options in given priority', () => {
  const options1 = Server.loadOptions({
    argv: ['', '--port', '31337'],
    env: { LOGUX_PORT: 21337 }
  }, { port: 11337 })
  const options2 = Server.loadOptions({
    argv: ['', '--port', '31337'],
    env: { LOGUX_PORT: 21337 }
  })
  const options3 = Server.loadOptions({
    argv: [],
    env: { LOGUX_PORT: 21337 }
  })

  expect(options1.port).toEqual(11337)
  expect(options2.port).toEqual(31337)
  expect(options3.port).toEqual(21337)
})

it('destroys everything on exit', () => checkOut('destroy.js'))

it('writes about unbind', () => {
  return check('unbind.js', [], 'kill').then(result => {
    expect(result[0]).toMatchSnapshot()
  })
})

it('shows uncatch errors', () => checkError('throw.js'))

it('shows uncatch rejects', () => checkError('uncatch.js'))

it('use environment variables for config', () => {
  process.env.LOGUX_PORT = 31337
  process.env.LOGUX_REPORTER = 'json'
  return checkOut('options.js')
})

it('uses reporter param', () => checkOut('options.js', ['', '--r', 'json']))

it('shows help', () => checkOut('options.js', ['', '--help']))

it('shows help about port in use', () => {
  return start('eaddrinuse.js').then(() => {
    return check('eaddrinuse.js')
  }).then(result => {
    expect(result[0]).toMatchSnapshot()
  })
})

it('shows help about privileged port', () => checkError('eacces.js'))

it('shows help about unknown option', () => checkError('unknown.js'))

it('shows help about missed option', () => checkError('missed.js'))

it('disables colors for constructor errors', () => {
  process.env.NODE_ENV = 'production'
  return checkError('missed.js')
})

it('uses reporter param for constructor errors', () => {
  return checkError('missed.js', ['', '--r', 'json'])
})

it('writes to bunyan log', () => checkOut('bunyan.js'))

it('writes to custom bunyan log', () => checkOut('bunyan-custom.js'))

it('writes using custom reporter', () => checkOut('custom-reporter.js'))
