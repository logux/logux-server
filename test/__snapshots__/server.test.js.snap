// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`destroys everything on exit 1`] = `
" INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.0.0
        Environment:  test
        Node ID:      server:FnXaqDxY
        Subprotocol:  1.0.0
        Supports:     1.x
        Listen:       ws://127.0.0.1:2000

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

 Custom destroy task finished
"
`;

exports[`disables colors for constructor errors 1`] = `
" FATAL  Missed client subprotocol requirements at 1970-01-01 00:00:00
        Check server constructor and Logux Server documentation

"
`;

exports[`shows help 1`] = `
"Options:
  --host, -h      Host to bind server.                                  [string]
  --port, -p      Port to bind server                                   [number]
  --key, -k       Path to SSL key                                       [string]
  --cert, -c      Path to SSL certificate                               [string]
  --reporter, -r  Reporter type              [string] [choices: \\"human\\", \\"json\\"]
  --help          Show help                                            [boolean]

Examples:
  test/servers/options.js --port 31337 --host 127.0.0.1
  LOGUX_PORT=1337 test/servers/options.js

Environment variables:
  LOGUX_HOST, LOGUX_PORT, LOGUX_KEY, LOGUX_CERT, LOGUX_REPORTER

"
`;

exports[`shows help about missed option 1`] = `
" FATAL  Missed client subprotocol requirements at 1970-01-01 00:00:00
        Check server constructor and Logux Server documentation

"
`;

exports[`shows help about port in use 1`] = `
" ERROR  Port :2001 already in use at 1970-01-01 00:00:00
        Another Logux server or other app already running on this port.
        Maybe you didn’t not stop server from other project or previous version
        of this server was not killed.

"
`;

exports[`shows help about privileged port 1`] = `
" ERROR  You are not allowed to run server on this port at 1970-01-01 00:00:00
        Try to change user (e.g. root) or use port >= 1024

"
`;

exports[`shows help about unknown option 1`] = `
" FATAL  Unknown option 'support' in server constructor at 1970-01-01 00:00:00
        Maybe there is a mistake in option name or this version of Logux Server
        doesn’t support this option

"
`;

exports[`shows uncatch errors 1`] = `
" FATAL  Test Error at 1970-01-01 00:00:00
        fake stacktrace

Error event: Test Error
 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`shows uncatch rejects 1`] = `
" FATAL  Test Error at 1970-01-01 00:00:00
        fake stacktrace

Error event: Test Error
 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`use environment variables for config 1`] = `
"{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"loguxServer\\":\\"0.0.0\\",\\"environment\\":\\"test\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:31337\\",\\"msg\\":\\"Logux server is listening\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"msg\\":\\"Shutting down Logux server\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`uses reporter param 1`] = `
"{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"loguxServer\\":\\"0.0.0\\",\\"environment\\":\\"test\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:1337\\",\\"msg\\":\\"Logux server is listening\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"msg\\":\\"Shutting down Logux server\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`uses reporter param for constructor errors 1`] = `
"{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":60,\\"err\\":{\\"message\\":\\"Missed client subprotocol requirements\\",\\"name\\":\\"Error\\"},\\"note\\":\\"Check server constructor and Logux Server documentation\\",\\"msg\\":\\"Missed client subprotocol requirements\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`writes about unbind 1`] = `
" INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`writes to bunyan log 1`] = `
"{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"loguxServer\\":\\"0.0.0\\",\\"environment\\":\\"test\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:2000\\",\\"msg\\":\\"Logux server is listening\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
{\\"name\\":\\"logux-server\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"msg\\":\\"Shutting down Logux server\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`writes to custom bunyan log 1`] = `
"{\\"name\\":\\"logux-server-custom\\",\\"customProp\\":\\"42\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"loguxServer\\":\\"0.0.0\\",\\"environment\\":\\"test\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:2000\\",\\"msg\\":\\"Logux server is listening\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
{\\"name\\":\\"logux-server-custom\\",\\"customProp\\":\\"42\\",\\"hostname\\":\\"localhost\\",\\"pid\\":21384,\\"level\\":30,\\"msg\\":\\"Shutting down Logux server\\",\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"v\\":0}
"
`;

exports[`writes using custom reporter 1`] = `
"Event: listen
Details: {\\"loguxServer\\":\\"0.0.0\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"environment\\":\\"test\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"server\\":false,\\"cert\\":false,\\"host\\":\\"127.0.0.1\\",\\"port\\":2000}
Event: destroy
Details: undefined
"
`;
